File: frontend/supabase/functions/_shared/cors.ts
function splitOrigins(value?: string | null): string[] {
  if (!value) return []
  return value
    .split(',')
    .map((o) => o.trim().replace(/\/$/, ''))
    .filter(Boolean)
}

function isProduction(): boolean {
  const env = (Deno.env.get('ENV') || Deno.env.get('APP_ENV') || Deno.env.get('NODE_ENV') || '').toLowerCase()
  // Default to production-safe behavior if unset/unknown.
  return env !== 'development' && env !== 'dev' && env !== 'local' && env !== 'test'
}

function isLocalhostOrigin(origin: string): boolean {
  return /^https?:\/\/(localhost|127\.0\.0\.1)(:\d+)?$/.test(origin)
}

export function corsHeaders(req: Request): Record<string, string> {
  const origin = req.headers.get('origin')?.replace(/\/$/, '')
  if (!origin) return {}

  const allowed = splitOrigins(Deno.env.get('APP_ORIGINS'))
  const allowLocalhost = !isProduction() && (Deno.env.get('ENV') || Deno.env.get('APP_ENV') || Deno.env.get('NODE_ENV'))

  const isAllowed = allowed.includes(origin) || (allowLocalhost && isLocalhostOrigin(origin))
  if (!isAllowed) return {}

  return {
    'Access-Control-Allow-Origin': origin,
    'Access-Control-Allow-Headers': 'authorization, x-client-info, apikey, content-type',
    'Access-Control-Allow-Methods': 'POST, GET, OPTIONS, PUT, DELETE',
    Vary: 'Origin',
  }
}

File: backend/app/cors.py
import os
from typing import List, Optional

from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware


def _split_origins(value: Optional[str]) -> List[str]:
    if not value:
        return []
    return [o.strip().rstrip("/") for o in value.split(",") if o.strip()]


def is_production_env() -> bool:
    # Common environment markers across platforms
    env = (os.getenv("ENV") or os.getenv("APP_ENV") or os.getenv("NODE_ENV") or "").lower()
    # Default to production-safe behavior if unset/unknown to avoid accidentally allowing localhost in prod.
    return env not in {"dev", "development", "local", "test"}


def allowed_origins() -> List[str]:
    # Comma-separated list of allowed origins, e.g.:
    # APP_ORIGINS="https://evater.com,https://www.evater.com,https://staging.evater.com"
    configured = _split_origins(os.getenv("APP_ORIGINS"))
    if configured:
        return configured
    if is_production_env():
        # Keep the deployed Evater frontend usable even when the existing
        # Modal secret has not yet gained APP_ORIGINS. Operators can still
        # override this list explicitly for another deployment domain.
        return ["https://evater.xyz", "https://www.evater.xyz"]
    return []


def allowed_origin_regex() -> Optional[str]:
    # Safe local-dev default without opening production. In prod, require explicit origins.
    if is_production_env():
        return None
    return r"^https?://(localhost|127\.0\.0\.1)(:\d+)?$"


def add_cors_middleware(app: FastAPI) -> None:
    app.add_middleware(
        CORSMiddleware,
        allow_origins=allowed_origins(),
        allow_origin_regex=allowed_origin_regex(),
        allow_credentials=True,
        allow_methods=["GET", "POST", "OPTIONS"],
        # Include common headers used by Supabase clients as well.
        allow_headers=["Authorization", "Content-Type", "apikey", "x-client-info"],
    )
